---
title: "Haplotyep frequencies"
author: "Reuben M. Yaa"
date: "01/11/2022"
output: html_document
---

```{r setup, include=TRUE}
root.dir <- here::here()
knitr::opts_chunk$set(echo = TRUE, root.dir=root.dir)
knitr::opts_knit$set(root.dir = root.dir)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

### Introduction

Fetching the sequences from alignments in PHYLIP format and storing them as list of objects. Using phyclust, the sequences are grouped into haplotypes.The frequencies of each haplotype in the total data is determined and presented as a plot with truncation of most significant haplotypes.

The sequences of the haplotypes are extracted for further analysis

#### --Exctraction of haplotypes
```{r results='hide'}
library (phyclust)
my.snp <- read.phylip("data/MSP-1_19.phy", code.type ="NUCLEOTIDE")
ret <- haplo.post.prob(my.snp$org, ploidy = 1)

#updated the haplo.post.prob() code using from  (get.medBB.skip.fun, edit = T)>, from github.com/snoweye/phyclust/issues/7 I raised

```

#### --Haplotype frequencies
```{r results=FALSE}
haplo_freq<-getcut.fun(sort(ret$haplo$hap.prob, decreasing =TRUE), 
                     nn = my.snp$nseq, plot = 1)
pdf("figures/haplotype_frequency.pdf", width = 13, height = 9, pointsize=18)#save as pdf file
haplo_freq
# The getcut.fun truncates the  significant haplotypes and the plots a graph 
dev.off()

```
#### --Haplotype DNA sequences

Haplotypes sequences  translated to amino -acid sequence in MEGA, and haplotypes renamed and new data file *haplotype_amino.fas* saved for next analysis

```{r}
write.fasta(ret$haplo$haplotype, "data/haplotypes.fas")

```

Figure available [haplotypes](https://github.com/mangiruben/Haplotype-population-analysis-of-pf-MSP1-in-Malaria/blob/main/figures/haplotype_frequency.pdf)
<hr>